Fixing tests which attempted to restart the Surefire booter JVM

jglick opened the pull request:
[Recent builds](https://ci.jenkins.io/job/Core/job/jenkins/job/master/1177/execution/node/67/log/?consoleFull) seem to be running this test over and over, finally timing out. I want to know if there is something about this test, or if there is some infrastructure change, etc. (I have not been able to track the issue down to a specific source commit.)

jglick committed:
Seeing if this helps.

batmat commented:
@jglick seems like the first run succeeded. I just closed/reopened to retrigger and see if this behaves differently.

How about use the way I used in #3797 and run *only* this test multiple times and see what it gives?

jglick commented:
Well build 1 passed indeed, but wound up running `JenkinsBuildsAndWorkspacesDirectoriesTest` a dozen times. So something is weird with the Surefire runner it seems.

jglick committed:
Noticed a Surefire error.
https://ci.jenkins.io/job/Core/job/jenkins/job/master/1179/execution/node/38/log/?consoleFull
[WARNING] Corrupted stdin stream in forked JVM 2. See the dump file /home/jenkins/workspace/Core_jenkins_master-CZB5BLBJFXZE2BKR63DSXQKXRZJAOWBX73QP55GU5KMX7AHAF6CQ/test/target/surefire-reports/2018-12-12T10-57-04_077-jvmRun2.dumpstream

jglick commented on Jenkinsfile:
```diff
@@ -61,6 +61,7 @@ for(j = 0; j < jdks.size(); j++) {
                 stage("${buildType} Publishing") {
                     if (runTests) {
                         junit healthScaleFactor: 20.0, testResults: '*/target/surefire-reports/*.xml'
+                        archiveArtifacts allowEmptyArchive: true, artifacts: '**/target/surefire-reports/*.dumpstream'
```
→ https://github.com/jenkinsci/remoting/pull/315

batmat replied:
interesting, wasn''t aware of this feature. NIT then reading https://maven.apache.org/surefire/maven-surefire-plugin/faq.html#dumpfiles don't we actually want *.dump* in case?

jglick replied:
I am not sure what `*.dump` is exactly, but anyway an actual `*.dumpstream` file led me to a source code issue. Can always improve as needed.

jglick committed:
JenkinsBuildsAndWorkspacesDirectoriesTest also has a restart issue, it seems.

jglick committed:
Merge branch 'master' into timeout-diagnosis

jglick committed:
Let us again try all matrix variants.

jglick commented:
> I think I know this could be fixed.

Hmm, my idea was `reuseForks=false` as in https://github.com/jenkinsci/workflow-cps-plugin/pull/178, but `test-pom` already does that!

jglick commented:
I cannot reproduce the error noted in #3772 running that _single_ test, but weird behavior appears once you start running other tests too:

```sh
mvn -f test surefire:test -Dtest=DisablePluginCommandTest,JenkinsBuildsAndWorkspacesDirectoriesTest,NodesTest,TransientActionFactoryTest -Pall-tests
```

```
[INFO] Running hudson.cli.DisablePluginCommandTest
[INFO] Running jenkins.model.JenkinsBuildsAndWorkspacesDirectoriesTest
[INFO] Running hudson.cli.DisablePluginCommandTest
..
[INFO] Results:
[INFO] 
[WARNING] Flakes: 
[WARNING] hudson.cli.DisablePluginCommandTest.canDisableDependentPluginWrongOrderStrategyAll(hudson.cli.DisablePluginCommandTest)
[INFO]   Run 1: PASS
[ERROR]   Run 2: DisablePluginCommandTest.canDisableDependentPluginWrongOrderStrategyAll
[INFO]   Run 3: PASS
[INFO] 
[WARNING] hudson.cli.DisablePluginCommandTest.disablePluginsStrategyAll(hudson.cli.DisablePluginCommandTest)
[ERROR]   Run 1: DisablePluginCommandTest.disablePluginsStrategyAll » Reactor java.lang.Error: ...
[INFO]   Run 2: PASS
[INFO]   Run 3: PASS
[INFO] 
[WARNING] jenkins.model.JenkinsBuildsAndWorkspacesDirectoriesTest.externalBuildDirectorySymlinks(jenkins.model.JenkinsBuildsAndWorkspacesDirectoriesTest)
[ERROR]   Run 1: JenkinsBuildsAndWorkspacesDirectoriesTest.externalBuildDirectorySymlinks
[INFO]   Run 2: PASS
[INFO] 
[INFO] 
[WARNING] Tests run: 65, Failures: 0, Errors: 0, Skipped: 0, Flakes: 3
```

jglick commented:
I think the problem is that the test JVM is getting a `UnixLifecycle`. Using `RestartableJenkinsRule` is fine since the restarts are purely in application code.

jglick committed:
Blocking Lifecycle.restart from within JenkinsRule.

jglick commented:
As an aside, I just noticed that https://github.com/jenkinsci/jenkins/commit/a59814e85ccba57c352b95b74128c96dbd723c24 puts the burden of calling `cleanUp` on each `Lifecycle` implementation, which seems silly—you would expect this to be part of `Jenkins.restart/safeRestart`.

MRamonLeon reviewed (approved):
Do you mean use `RestartableJenkinsRule` in the `DisablePluginCommandTest` and remove the _Ignores_ @jglick? I didn't use it because `EnablePluginCommandTest` was so and I didn't need to do things on Jenkins after the restart.

jglick commented:
> Do you mean use `RestartableJenkinsRule` in the `DisablePluginCommandTest`

No, because `RestartableJenkinsRule` is used when the restart is initiated by the test directly (by means of separating it into blocks), whereas these tests are calling some code which in turn calls `Jenkins.restart` outside of the test’s direct control.

jglick committed:
Fixing a SIOOBE encountered during Windows-jdk11 tests (but not reproducible locally).
Amends code added in https://github.com/jenkinsci/jenkins/pull/2082.
java.lang.RuntimeException: Unexpected issues encountered during cleanUp: offset 0, count -1, length 5
	at jenkins.model.Jenkins.cleanUp(Jenkins.java:3319)
	at org.jvnet.hudson.test.JenkinsRule.after(JenkinsRule.java:495)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:565)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.StringIndexOutOfBoundsException: offset 0, count -1, length 5
	at java.base/java.lang.String.checkBoundsOffCount(String.java:3304)
	at java.base/java.lang.String.<init>(String.java:465)
	at hudson.TcpSlaveAgentListener$PingAgentProtocol.connect(TcpSlaveAgentListener.java:416)
	at hudson.TcpSlaveAgentListener.shutdown(TcpSlaveAgentListener.java:196)
	at jenkins.model.Jenkins._cleanUpShutdownTcpSlaveAgent(Jenkins.java:3548)
	at jenkins.model.Jenkins.cleanUp(Jenkins.java:3294)
	... 6 more

batmat commented on core/src/main/java/hudson/TcpSlaveAgentListener.java:
```diff
@@ -415,7 +415,9 @@ public boolean connect(Socket socket) throws IOException {
                                     socket.getRemoteSocketAddress(),
                                     new String(ping, "UTF-8"),
-                                    new String(response, 0, responseLength, "UTF-8")
+                                    responseLength > 0 && responseLength <= response.length ?
```
FTR this fix is for `UsageStatisticsTest` last failure:

![image](https://user-images.githubusercontent.com/223853/49905076-a7cdb700-fe6c-11e8-84f6-323f098fa87a.png)

batmat reviewed (approved):
LGTM, let's merge once we get a green build.

I'll see to be if we can analyze all these through https://issues.jenkins-ci.org/browse/JENKINS-55122 once we get the situation back in control.

batmat commented:
@LinuxSuRen in short, if you need that very fix, there's none.
Where are you seeing it? Thanks

LinuxSuRen commented:
During test junit test of all my plugins. I wonder that shall we should fix this in the older LTS? @batmat

batmat commented:
You mean when running your plugin's builds using Java 11? Sorry I am trying to make sure I get it.

Anyway if you have a case for backporting this, this will need a JIRA labelled 'lts-candidate' and explanarions as to why it would be backported.

Cheers

batmat commented:
Does it also fail on the command line using `mvn test` then? It will help
know if it's a general issue or specific to IntelliJ.

Le dim. 30 déc. 2018 à 05:25, Zhao Xiaojie <notifications@github.com> a
écrit :

> I run my junit test from the IDEA. Java version is 8.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/3804#issuecomment-450538417>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANqbSYQqGkL7i9G-4J0owv7Y-n9pB1zks5u-EBGgaJpZM4ZOqmv>
> .
>

LinuxSuRen commented:
They're the same results. You can test it by using my repo. The link is here https://github.com/LinuxSuRen/alauda-devops-pipeline-plugin/tree/add-test-case.

batmat commented:
@LinuxSuRen the failure I see locally seems totally unrelated, just the typical issue with a wrong version of a plugin that prevents Jenkins from starting up.

I would recommend the following:

* let's stop commenting here as it creates noise for people subcribed here, and if anyway related we should discuss this in the PR, see the next point
* file https://github.com/LinuxSuRen/alauda-devops-pipeline-plugin/tree/add-test-case as a PR against https://github.com/jenkinsci/alauda-devops-pipeline-plugin/ so we can check how it behaves there.
** We'll be able to use comments there to discuss your specific issue
** You really should update to the latest parent pom, this issue I'm seeing is now impossible with latest plugin parent pom with `RequireUpperBoundDeps` rule enabled.

Thanks
